src/lib/uptime_service/uptime_service.ml

```diff
@@ -0,0 +1,431 @@
+open Signature_lib
+
+module Blake2 = Blake2.Make ()
+
+type block_data =
+  { block: string
+  ; created_at: string
+  ; peer_id: string
+  ; snark_work: string option [@default None] }
+[@@deriving to_yojson]
+
+type proof_data = Ledger_proof.Stable.Latest.t * Core_kernel.Time.Span.t
+[@@deriving bin_io_unversioned]
+
+(* TODO: what's the max size of a serialized block? *)
+let block_buf = Bin_prot.Common.create_buf 12_000_000
+
+(* TODO: what the size of a serialized proof? *)
+
+let proof_buf = Bin_prot.Common.create_buf 500_000
+
+let external_transition_of_breadcrumb breadcrumb =
+  let {With_hash.data= external_transition; _}, _ =
+    Transition_frontier.Breadcrumb.validated_transition breadcrumb
+    |> External_transition.Validated.erase
+  in
+  external_transition
+
+let get_rfc3339_time () =
+  let tm = Unix.gettimeofday () in
+  match Ptime.of_float_s tm with
+  | None ->
+      (* should never occur *)
+      failwith "In uptime service, could not convert current time to Ptime.t"
+  | Some ptime ->
+      Ptime.to_rfc3339 ~tz_offset_s:0 ptime
+
+let send_uptime_data ~logger ~(submitter_keypair : Keypair.t) ~url ~state_hash
+    block_data =
+  let block_data_json = block_data_to_yojson block_data in
+  let block_data_string = Yojson.Safe.to_string block_data_json in
+  let signature =
+    String_sign.Schnorr.sign submitter_keypair.private_key block_data_string
+  in
+  let json =
+    (* JSON structure in issue 9110 *)
+    `Assoc
+      [ ("data", block_data_json)
+      ; ("signature", Signature.to_yojson signature)
+      ; ("submitter", Public_key.to_yojson submitter_keypair.public_key) ]
+  in
+  let headers = Cohttp.Header.of_list [("Content-Type", "application/json")] in
+  match%map
+    Monitor.try_with ~here:[%here] ~extract_exn:true (fun () ->
+        match%map
+          Cohttp_async.Client.post ~headers
+            ~body:(Yojson.Safe.to_string json |> Cohttp_async.Body.of_string)
+            url
+        with
+        | {status; _}, _body ->
+            if Cohttp.Code.code_of_status status = 200 then
+              [%log info]
+                "Sent block with state hash $state_hash to uptime server at \
+                 URL $url"
+                ~metadata:
+                  [ ("state_hash", State_hash.to_yojson state_hash)
+                  ; ( "includes_snark_work"
+                    , `Bool (Option.is_some block_data.snark_work) )
+                  ; ("url", `String (Uri.to_string url)) ]
+            else
+              [%log error]
+                "Failure when sending block with state hash $state_hash to \
+                 uptime server at URL $url"
+                ~metadata:
+                  [ ("state_hash", State_hash.to_yojson state_hash)
+                  ; ("url", `String (Uri.to_string url))
+                  ; ("http_code", `Int (Cohttp.Code.code_of_status status))
+                  ; ( "http_error"
+                    , `String (Cohttp.Code.string_of_status status) ) ] )
```

Comment:
I suggest including json body that was posted to support BP being able to prove (and be ~trusted) that work was submitted. I suppose that the signed data could still be faked after the fact, so maybe not worth it / not any more trustworthy than a reported HTTP error? It would prevent any trivial disputes though.

I suspect we are already in an environment where there are >120 BP with perfect uptime, so missing a single report could be enough to bump someone from the list. Logging the body on error would seem to be worthwhile - to me.